### ISSUE TITLE
Support activeBuildProfile parameter

### ISSUE THREAD

**Issue by Sammmte** [game-ci/unity-builder#674]
**Context**

When using "activeBuildProfile" parameter (in "customParameters", example: `customParameters: -activeBuildProfile Assets/Settings/BuildProfiles/WindowsDevelopment.asset`) [(for Unity 6)](https://docs.unity3d.com/6000.0/Documentation/Manual/EditorCommandLineArguments.html) builder action throws an error message:

```
Aborting batchmode due to failure:
Can't have -activeBuildProfile argument defined when using -buildTarget argument
```

Problem is "targetPlatform" is a required parameter.

**Suggested solution**

Maybe require to have either "targetPlatform" parameter OR another parameter like "buildProfile". "buildProfile" parameter would hold the same value "activeBuildProfile" requires, a relative path to the profile asset.

**Comment by GabLeRoux** [comment]
Yeah Build Profiles is quite a recent feature which is indeed something that would be worth adding as a feature 👍 

[More details about the build profiles](https://docs.unity3d.com/6000.0/Documentation/Manual/build-profiles.html)

![image](https://github.com/user-attachments/assets/6b041589-0a94-4e8c-8adb-f4c1d09dd326)

**Comment by OwenSignaturize** [comment]
This would be incredibly useful

**Comment by GabLeRoux** [comment]
@coderabbitai read the following files and suggest an approach to add support for activeBuildProfile parameter for unity version >= 6 (6000)

dist/default-build-script/Assets/Editor/UnityBuilderAction/Builder.cs
dist/default-build-script/Assets/Editor/UnityBuilderAction/Versioning/GitException.cs
src/index.ts
src/model/platform-setup/setup-windows.ts
src/model/unity.test.ts
src/model/input.ts
dist/default-build-script/Assets/Editor/UnityBuilderAction/Versioning/VersionApplicator.cs
src/model/action.ts
dist/default-build-script/Assets/Editor/UnityBuilderAction/Versioning/Git.cs
dist/default-build-script/Assets/Editor/UnityBuilderAction/Versioning/VersionGenerator.cs
src/model/action.test.ts
src/model/unity.ts
dist/default-build-script/Assets/Editor/UnityBuilderAction/Input/ArgumentsParser.cs
dist/default-build-script/Assets/Editor/UnityBuilderAction/Reporting/StdOutReporter.cs
dist/default-build-script/Assets/Editor/UnityBuilderAction/Reporting/CompileListener.cs
src/model/input.test.ts
src/model/platform-setup/setup-mac.ts
src/model/platform-setup/setup-android.ts
src/model/docker.ts
src/model/docker.test.ts
action.yml

**Comment by GabLeRoux** [comment]
@coderabbitai There is an existing command line argument for this on the unity cli, it is `-activeBuildProfile`. Here is an example of how I used it in a different project (this is a confirmed to be working solution used with GitlabCI, not the GameCI Unity Builder GitHub Action):

```shell
#!/usr/bin/env bash

set -euo pipefail

echo "Building for '$BUILD_TARGET'"

# Ensure the build path is defined and exists
export BUILD_PATH="$UNITY_DIR/Builds/$BUILD_TARGET/"
mkdir -p "$BUILD_PATH"

# Base Unity command split into an array
mapfile -t UNITY_CMD < <(
  printf "%s\n" \
    "${UNITY_EXECUTABLE:-xvfb-run}" \
    "--auto-servernum" \
    "--server-args=-screen 0 640x480x24" \
    "unity-editor"
)

# Base Unity arguments
mapfile -t UNITY_ARGS < <(
  printf "%s\n" \
    "-projectPath" "$UNITY_DIR" \
    "-quit" \
    "-batchmode" \
    "-nographics" \
    "-customBuildTarget" "$BUILD_TARGET" \
    "-customBuildName" "$BUILD_NAME" \
    "-customBuildPath" "$BUILD_PATH" \
    "-executeMethod" "BuildCommand.PerformBuild" \
    "-logFile" "/dev/stdout"
)

# Adjust arguments based on whether a build profile is provided
if [[ -n "${BUILD_PROFILE_PATH:-}" && -f "$BUILD_PROFILE_PATH" ]]; then
  echo "Using build profile: '$BUILD_PROFILE_PATH'"
  UNITY_ARGS+=("-activeBuildProfile" "$BUILD_PROFILE_PATH")
else
  echo "No build profile specified or the file does not exist; proceeding with default build target."
  UNITY_ARGS+=("-buildTarget" "$BUILD_TARGET")
fi

# Print arguments for debugging
echo "Unity Command: ${UNITY_CMD[*]}"
echo "Unity Arguments: ${UNITY_ARGS[*]}"

# Run the Unity command with arguments
"${UNITY_CMD[@]}" "${UNITY_ARGS[@]}"

UNITY_EXIT_CODE=$?

# Check the Unity exit code and display appropriate messages
if [[ $UNITY_EXIT_CODE -eq 0 ]]; then
  echo "Build succeeded with no failures."
elif [[ $UNITY_EXIT_CODE -eq 2 ]]; then
  echo "Build succeeded, but some tests failed."
elif [[ $UNITY_EXIT_CODE -eq 3 ]]; then
  echo "Build failed due to other reasons."
else
  echo "Unexpected Unity exit code: $UNITY_EXIT_CODE."
fi

# List the build output directory contents
ls -la "$BUILD_PATH"

# Fail the job if the build folder is empty
if [[ -z "$(ls -A "$BUILD_PATH")" ]]; then
  echo "Build folder is empty! Failing the job."
  exit 1
fi
```

Here are some additional details I am aware of:

1. It is important to not pass a full path or even a relative path starting with `./` to the `-activeBuildProfile` parameter. It must use a local path to the asset file. For exmaple:
```
BUILD_PROFILE_PATH: "Assets/Settings/Build Profiles/Android Mobile.asset"
BUILD_PROFILE_PATH: "Assets/Settings/Build Profiles/Android VR Oculus.asset"
```
2. The documentation for the `-activeBuildProfile` is available here: https://docs.unity3d.com/Manual/EditorCommandLineArguments.html


| Command                          | Description                                                                                                                                                                                                                                                                                                                                                                  |
|----------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `-activeBuildProfile <pathname>` | Set the build profile saved at the given path as an active build profile (e.g., `-activeBuildProfile "Assets/Settings/Build Profiles/WindowsDemo.asset"`). Applies custom scripting defines found in the Build data section of the active build profile before compilation. These scripting defines are additive and don’t override other scripting defines in your project. |

**Comment by webbertakken** [comment]
Impressive @GabLeRoux :)

**Comment by webbertakken** [comment]
Ended up hiding those comments as they add a stream of information that isn't really helpful.

For whoever feels like taking a stab at this, please have a look at the combination of the first comment from Coderabbat and the followup from @GabLeRoux.

**Comment by JanikHelbig-NB** [comment]
I've looked a bit at this already and so far the biggest issue (I think) has not been mentioned.
(Which is why I think that AI assistants are mostly useless, surface-level noise; but whatever)

The action currently does a bit of build platform specific setup depending on the build target. If the build profile were to be specified instead of the build target, the necessary information would not be available at the required time.

We could try to extract the build target information from the BuildProfile asset file directly and use that to fill in the action parameter. That way we wouldn't need to change most of the existing code. I'm just not sure how robust and reliable that would be.

**Comment by webbertakken** [comment]
We're just casually trialing this bot. It's been doing some very helpful comments as well.

Agreed that auto-detect would be preferable where possible. We already do that for the `unityVersion` or `editorVersion` param and could do the same for this.

That said, I think the whole point of Build Profiles is to have multiple. So beyond auto-detecting we do need to be able to specify it as a param as well.

**Comment by MichaelBuhler** [comment]
My team has run into this problem as well, and I'd like to take a crack at resolving this issue.

-----

@JanikHelbig-NB is right when he says "The action currently does a bit of build platform specific setup depending on the build target".

However, I don't know that auto-detection is preferable. The `targetPlatform` GitHub action parameter is _not_ necessarily a one-to-one mapping to the Unity `-buildTarget` CLI option. It is also an instruction to this `unity-builder` GitHub action of which image to pull and run from Docker Hub.

Therefore, we should be able to simply add an optional `buildProfile` GitHub action parameter to this GitHub action, keeping the required `targetPlatform` parameter. Then, immediately before the command to run the Unity build (i.e. in `build.sh` or `build.ps1`), the script decides whether it will pass `-buildTarget ...` or `-activeBuildProfile ...` in as a CLI param, based on the presence of the (new) `BUILD_PROFILE` env var, which will be set from the `buildProfile` GitHub action param.

In this case, if the end user (GitHub workflow author) instructs the provides incompatible values for the `targetPlatform` and `buildProfile` GitHub action params (for instance, a Windows target platform with a WebGL build profile), then Unity will throw an error when it runs, because the required build platform packages are not installed via Unity Hub.

I can quickly pursue this path.

-----

Alternately, I can try the auto detection. A core difference here is that the current `targetPlatform` GitHub action parameter would become optional, as well as adding the new, and also optional, `buildProfile` parameter. Logic will need to be added to disallow providing both parameters, or first check the build profile to first see if it is compatible with target platform before continuing. Alternately, we can document that if both parameters are provided, the `targetPlatform` is ignored.

In this case of auto-detection, the GitHub action will need to open the provided build profile `.asset` file, parse the yaml, and inspect `MonoBehavior.m_PlatformBuildProfile.rid`. It will use that `rid` find the appropriate entry within `MonoBehavior.references.RefIds`. Within that reference entry, the `type` (`type.class` or `type.ns`) should have what we need to run the correct Docker image.

The build script (`build.sh` or `build.ps1`) will still need to choose whether to pass `-buildTarget ...` or `-activeBuildProfile ...` into Unity, as these are mutually exclusive options.

-----

@webbertakken Let me know your thoughts.

**Comment by davidmfinol** [comment]
I agree that it would be better to go with the first option and simply add an optional `buildProfile` parameter to unity-builder.

We could read the build profile's `.asset` file to get the platform and therefore allow `targetPlatform` to be optional, but I think that's unnecessary for the first iteration of support for build profiles.

In general, I think it's better to be iterative with smaller updates, and we can follow up with making target platform optional after we've got the initial support for `buildProfile`.

For context, I'm also planning on using build profiles for new builds of my personal project, so that's also part of the reason I'm interested in quicker iterations for this.

**Comment by webbertakken** [comment]
Yea, sounds good to me. 

As long as we can stay backwards compatible and people don't have to change their pipelines just to upgrade the action (i.e. only need changes if they opt-in to `buildProfile`).

**Comment by MichaelBuhler** [comment]
@GabLeRoux @webbertakken @davidmfinol I have opened draft PRs here:

- https://github.com/game-ci/unity-builder/pull/685
- https://github.com/game-ci/documentation/pull/498

which do as requested: add a new, optional `buildProfile` parameter to `unity-builder`. As I said, the build script decides whether to pass `-buildTarget` OR `-activeBuildProfile` into Unity based on the presence of the new `buildProfile` parameter. This should be backwards compatible with existing consumers.

I have left the PRs in draft right now, as I am not really sure if or how I will be able to test Mac and Windows builds. Furthermore, my workflows on GitHub are failing (on `docker pull`) due to the default GitHub Actions runner running out of disk space, which I have not worked around yet. Let me know if you are able to assist me with testing.

```
docker: failed to register layer: write /opt/unity/Editor/Data/Tools/netcorerun/System.Private.Xml.dll: no space left on device.
```

**Comment by MichaelBuhler** [comment]
Perhaps something has changed since @Sammmte's original post four months ago. When I run Unity with a `-activeBuildProfile ...` I get an error `Missing argument -buildTarget` and the process exits. Maybe these are no longer mutually exclusive?

**Comment by MichaelBuhler** [comment]
Things got more complicated than I expected. This `unity-builder` repo uses a custom build method.

I can pass `-projectPath`, `-activeBuildProfile`, and `-build` into Unity and get a successful, headless build locally:

```shell
/Applications/Unity/Hub/Editor/6000.0.33f1/Unity.app/Contents/MacOS/Unity \
  -logFile - \
  -quit \
  -batchmode \
  -nographics \
  -projectPath "/absolute/path/to/unity/project" \
  -activeBuildProfile "relative/path/to/build-profile.asset" \
  -build "/absolute/path/to/build/output/folder"
```

The problem is that `unity-builder` _always_ passes `-executeMethod` into Unity; it _never_ uses the default batch mode method.

So, if we don't want to require the user to define a custom static build class method to use build profiles with this GitHub action, I'll have to also modify the [`dist/default-build-script/*`](https://github.com/game-ci/unity-builder/tree/83c85328dd491a4268b41f1aae9de279e1fe7be5/dist/default-build-script/Assets/Editor/UnityBuilderAction) within this repo.

-----

It was a very long time ago (5years 1month) when the decision was made to always execute the `unity-builder` custom build method: https://github.com/game-ci/unity-builder/commit/5dcd509017c77a4ac38dcbb331b45d2b43950edb#diff-6f9d41d046756f0ddc2fcee0626bdb50100d12b88f293734eff742818e03efa2R68

@webbertakken What was the reasoning there? I wonder if it was a way to provide a mechanism to programmatically parameterize/customize more build options--the exact functionality that has now been added to Unity as "Build Profiles".

Some thought could be given to adding (yet) another GitHub action param, such as `noBuildMethod: true`, which would skip `unity-builder`'s custom build method.

Another option is that if the user provides the `buildProfile` GitHub action param, and does _not_ provide the `executeMethod` param, then Unity is _not_ passed `-executeMethod UnityBuilderAction.Builder.BuildProject` on the CLI, though I suspect this will have other negative consequences.

**Comment by MichaelBuhler** [comment]
I have successfully built locally with a build profile and a modified version of `UnityBuilderAction.Builder.BuildProject()`, visible here https://github.com/MichaelBuhler/unity-builder/commit/222b8068bdfe2da016b71418908862891db99cdd, summarized here:

```csharp
      if (options.ContainsKey("buildProfile")) {

#if UNITY_6000_0_OR_NEWER
        // Load build profile from Assets folder
        BuildProfile buildProfile = AssetDatabase.LoadAssetAtPath<BuildProfile>(options["buildProfile"]);

        // Set it as active
        BuildProfile.SetActiveBuildProfile(buildProfile);

        // Define BuildPlayerWithProfileOptions
        buildPlayerOptions = new BuildPlayerWithProfileOptions {
            buildProfile = buildProfile,
            locationPathName = options["customBuildPath"],
            options = buildOptions,
        };
#endif

      } else {
        // old code here
      }
```

**Comment by MichaelBuhler** [comment]
> One question: Will there be cases where buildProfiles will completely remove the need for customBuildMethod?

@webbertakken, yes that it currently the case.

The name of the "customBuildMethod" is passed to Unity via the `-executeMethod ...` CLI option. I can build my project without passing `-executeMethod` at all; I only need to supply the `-projectPath`, `-activeBuildProfile`, and `-build` options to successfully run a build.

**Comment by webbertakken** [comment]
@MichaelBuhler ok, then would the features from the build script still work? I.e. do we need the BuildProject method at all in case ˋbuildProfileˋ is passed?

**Comment by MichaelBuhler** [comment]
@webbertakken It is not _necessary_ to do a Unity build with a custom method, such as `BuildProject()`. The current [code changes that I have drafted](https://github.com/game-ci/unity-builder/compare/83c85328dd491a4268b41f1aae9de279e1fe7be5...222b8068bdfe2da016b71418908862891db99cdd) still use the `BuildProject()` custom build method.

It would be possible for me to rework it (continue iterating) to provide a way bypass the use `BuildProject()` entirely.

From what I can tell, the `BuildProject()` custom build method mainly provides features to determine/set the application version and, in the case of an Android build, to set Android parameters.

**Comment by MichaelBuhler** [comment]
@webbertakken @davidmfinol @GabLeRoux

I made some simplifications, tested with a self-hosted (macOS) GitHub Actions Runner (n.b. this uses the [MacBuilder](https://github.com/game-ci/unity-builder/blob/main/src/model/mac-builder.ts#L5) to run Unity directly--not from a `unityci/editor` Docker Hub image), and submitted my PR for review: #685 

I also added two more matrix tests to the Ubuntu tests, which will test Unity 6 builds both with and without a Build Profile.

**Comment by webbertakken** [comment]
Thanks for explaining.

> It would be possible for me to rework it (continue iterating) to provide a way bypass the use BuildProject() entirely.

I think it's not needed to do immediately. There may be advantages to sticking to the custom build method for now, as we learn more.

**Comment by ryo0ka** [comment]
@webbertakken Hi, excuse me to necro this ticket. 

Our team has a specific use case that your current workaround doesn't cover and we do want to have a proper interface to `-activeBuildProfile`. In short, our build relies on preprocessors that are listed in each build profile in order to link platform-specific assemblies. 

`BuildProfile.SetActiveBuildProfile()` (as in @MichaelBuhler 's suggestion) won't immediately take effect for preprocessors: it will "queue" the script recompile for the next editor update, which doesn't take place in batch mode.

AFAIK before 6000+, this kind of ci/cd scenarios needed to run two executeMethod passes which is partly why a lot of teams opted out to self-hosted runners with Unity Editor pre-installed.

Maybe I can tackle on implementation if you guys don't mind. Also I'm all ears for other workarounds. cheers

**Comment by webbertakken** [comment]
@ryo0ka thank you for your comment.

Sure, I think many would be happy if you would like to take a stab at this. 

There is a fair amount of information already in this thread. And multiple approaches. Would you mind opening a new ticket, slightly better outlining your use case (as I didn't understand it word for word) as well as the proposed solution?

**Comment by timcassell** [comment]
> it will "queue" the script recompile for the next editor update, which doesn't take place in batch mode.

I believe it can be made to work by not passing `-quit` argument, and quitting manually in a script. The unity-test-runner does this for Standalone tests.

**Comment by ryo0ka** [comment]
@webbertakken will do:)
@timcassell Hi, that's great to know. Thank you for the heads up :) Is it possible to not pass `-quit` to unity from game ci?
